### PR TITLE
Testing/forces update

### PR DESCRIPTION
**Pull request by shuds13**
Update forces tests.

For relevant forces tests
- [x] Use persistent generator as default
- [x] Use new OO interface

Other capabilities
- [x] Add generator that sets num_gpus based on x value in range(e.g. number of particles)
- [x] Update regression test test_GPU_variable_resources.py to use new gen (as well as old).

Tests
- [x] Updating forces_simple and forces_gpu and prob. forces_multi_task/
- [x] Remove forces_gpu_persis_gen - as now that's default
- [x] Add a var resources test (separate now - and using new num_gpus generator)

Update tutorials
- [x] Forces simple
- [x] Forces GPU

### CODE REVIEW

File: docs/tutorials/executor_forces_tutorial.rst [hunk 1]
```diff
@@ -35,97 +38,135 @@ Calling Script
...
 
 Registering an application is as easy as providing the full file-path and giving
 it a memorable name. This Executor will later be used within our simulation
 function to launch the registered app.
 
-Next define the :ref:`sim_specs<datastruct-sim-specs>` and
-:ref:`gen_specs<datastruct-gen-specs>` data structures. Recall that these
-are used to specify to libEnsemble what user functions and input/output fields to
-expect, and also to parameterize function instances without hard-coding:
+On line 22, we initialize the ensemble. The :meth:`parse_args()<tools.parse_args>`
```
Comment (jlnav) [inline docs/tutorials/executor_forces_tutorial.rst:84]: Perhaps this should link to or describe the `Ensemble` docs `parse_args=True` parameter? Something like:

"The `parse_args=True` parameter instructs the Ensemble class to read `comms` and `nworkers` from the command line."

File: docs/tutorials/executor_forces_tutorial.rst [hunk 2]
```diff
@@ -35,97 +38,135 @@ Calling Script
...
+is used to read `comms` and `nworkers` from the command line. This sets
+the respective `libE_specs` options.
+
+Next we add basic configuration for the ensemble. As one worker will run a persistent
+generator that will not need additinal computing resources, we calculate the number
+of workers that need resources to run simulations. We also set `sim_dirs_make`
+so that a directory is created for each simulation.  This helps organize output and
+also helps prevents workers from overwriting previous results.
```
Comment (jlnav) [inline docs/tutorials/executor_forces_tutorial.rst:92]: "helps prevent"

File: docs/tutorials/executor_forces_tutorial.rst [hunk 3]
```diff
@@ -35,97 +38,135 @@ Calling Script
...
 
 .. code-block:: python
-    :linenos:
+  :linenos:
+  :lineno-start: 38
+
+    ensemble.sim_specs = SimSpecs(
+        sim_f=run_forces,
+        inputs=["x"],
+        out=[("energy", float)],
```
Comment (jlnav) [inline docs/tutorials/executor_forces_tutorial.rst:118]: We can use `outputs=` now for symmetry

File: docs/tutorials/executor_forces_tutorial.rst [hunk 4]
```diff
@@ -134,48 +175,56 @@ This may take some additional browsing of the docs to complete.
 
 Write an alternative Calling Script similar to above, but with the following differences:
 
- 1. Add an additional worker directory so workers operate in ``/scratch/ensemble`` instead of the default current working directory.
+ 1. Set :ref:`libEnsemble's logger<logger_config>` to print debug messages.
  2. Override the MPIExecutor's detected MPI runner with ``"openmpi"``.
- 3. Set :ref:`libEnsemble's logger<logger_config>` to print debug messages.
- 4. Use the :meth:`save_libE_output()<tools.save_libE_output>` function to save the History array and ``persis_info`` to files after libEnsemble completes.
+ 3. Tell the allocation function to return results to the generator asychronously.
+ 4. Use the ensemble function :meth:`save_output()<ensemble.Ensemble.save_output>` to save the History array and ``persis_info`` to files after libEnsemble completes.
```
Comment (jlnav) [inline docs/tutorials/executor_forces_tutorial.rst:181]: Such paths need to be full, I think: `libensemble.ensemble.Ensemble.save_output`

**Review by jlnav — APPROVED** [review]
Looks good!

File: docs/tutorials/executor_forces_tutorial.rst [hunk 5]
```diff
@@ -35,97 +38,135 @@ Calling Script
...
 
-  # Instruct libEnsemble to exit after this many simulations
-  exit_criteria = {"sim_max": 8}
+    # Instruct libEnsemble to exit after this many simulations
+    ensemble.exit_criteria = ExitCriteria(sim_max=8)
 
-  # Seed random streams for each worker, particularly for gen_f
-  persis_info = add_unique_random_streams({}, nworkers + 1)
+    # Seed random streams for each worker, particularly for gen_f
+    ensemble.add_random_streams()
```
[the remark this answers is not in the document]
Comment (shuds13) [reply docs/tutorials/executor_forces_tutorial.rst:166]: Good point. I'm realising it means an extra import (from tools) which is a pain.

[the remark this answers is not in the document]
Comment (jlnav) [reply docs/tutorials/executor_forces_tutorial.rst:166]: That's one of the reasons I made `Ensemble.add_random_streams()` originally, I think I forgot to mention. Avoid an import, keep it simple for 75% of our tests

File: docs/tutorials/executor_forces_tutorial.rst (hunk 3 above)
Comment (shuds13) [reply docs/tutorials/executor_forces_tutorial.rst]: done

**Pull request merged**